Mathlib/Algebra/Order/Quantale.lean

```diff
@@ -0,0 +1,161 @@
+/-
+Copyright (c) 2024 Pieter Cuijpers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pieter Cuijpers
+-/
+import Mathlib.Algebra.Group.Defs
+import Mathlib.Algebra.Order.Monoid.Unbundled.Basic
+import Mathlib.Order.CompleteLattice
+
+/-!
+# Theory of quantales
+
+Quantales are the non-commutative generalization of locales/frames and as such are linked
+to point-free topology and order theory. Applications are found throughout logic,
+quantum mechanics, and computer science (see e.g. [Vickers1989] and [Mulvey1986]).
+
+The most general definition of quantale occurring in literature, is that a quantale is a semigroup
+distributing over a complete sup-semilattice. In our definition below, we use the fact that
+every complete sup-semilattice is in fact a complete lattice, and make constructs defined for
+those immediately available. Another view could be to define a quantale as a complete idempotent
+semiring, i.e. a complete semiring in which + and sup coincide. However, we will often encounter
+additive quantales, i.e. quantales in which the semigroup operator is thought of as addition, in
+which case the link with semirings will lead to confusion notationally.
+
+In this file, we follow the basic definition set out on the wikipedia page on quantales,
+using a mixin typeclass to make the special cases of unital, commutative, idempotent,
+integral, and involutive quantales easier to add on later.
+
+## Main definitions
+
+* `IsQuantale` and `IsAddQuantale` : Typeclass mixin for a (additive) semigroup distributing
+  over a complete lattice, i.e satisfying `x * (sSup s) = ⨆ y ∈ s, x * y` and
+  `(sSup s) * y = ⨆ x ∈ s, x * y`;
+
+* `leftMulResiduation`, `rightMulResiduation`, `leftAddResiduation`, `rightAddResiduation` :
+  Defining the left- and right- residuations of the semigroup (see notation below).
+
+* Finally, we provide basic distributivity laws for sSup into iSup and sup, monotonicity of
+  the semigroup operator, and basic equivalences for left- and right-residuation.
+
+## Notation
+
+* `x ⇨ₗ y` : `sSup {z | z * x ≤ y}`, the `leftMulResiduation` of `y` over `x`;
+
+* `x ⇨ᵣ y` : `sSup {z | x * z ≤ y}`, the `rightMulResiduation` of `y` over `x`;
+
+## References
+
+<https://en.wikipedia.org/wiki/Quantale>
+<https://encyclopediaofmath.org/wiki/Quantale>
+<https://ncatlab.org/nlab/show/quantale>
+
+-/
+
+open Function
+
+namespace Quantale
+
+/-- An additive quantale is an additive semigroup distributing over a complete lattice. -/
+class IsAddQuantale (α : Type*) [AddSemigroup α] [CompleteLattice α] where
```

Comment:
I suspect we don't want this definition in the `Quantale` namespace.

Reply:
Agree.

Took the chance to include your suggestion to use a typeclass alias `Quantale` for `Semigroup CompleteLattice IsQuantale`. I hope that does not disrupt the review process too much.

Was wondering whether the use of `variable?` in a library file is okay, or whether I should unfold that...